activate.go

```diff
@@ -615,65 +703,236 @@ func (m *activateOneContainerStateMachine) runNextState(ctx context.Context) err
+//     storage container is part of the same install as the first container. The recovered
+//     primary key can be used to check the binding of subsequent storage containers using
+//     the first method.
+//
+// If more than one container is unlocked using a recovery key, then all containers must be
+// unlocked using a recovery key. This is because it is only possible to prove the binding
+// of storage containers to a single container using a platform that is registered with the
+// PlatformProtectedByStorageContainer flag.
+//
+// Once any storage containers have been unlocked using a platform keyslot, all subsequent
+// containers must be unlocked using platform keyslots. This is because it is not possible
+// to properly verify that subsequent containers are bound to existing containers if they
+// are unlocked with a recovery key. In this case, existing containers must be deactivated
+// and then activated only with recovery keys.
+// XXX: A mechanism to handle this case will be provided in the future.
 func (c *ActivateContext) ActivateContainer(ctx context.Context, container StorageContainer, opts ...ActivateOption) error {
-	// Procss options. These apply on top of those supplied to NewActivateContext.
+	// Process options. These apply on top of those supplied to NewActivateContext.
 	cfg := c.cfg.Clone()
 	for _, opt := range opts {
 		opt.ApplyOptionToConfig(cfg)
 	}
 
-	// TODO: When we retain activation state, make sure we obtain the activation
-	// status and keyslot errors from the state machine before returning.
-
 	// Create and run a state machine for this activation
-	sm := newActivateOneContainerStateMachine(container, cfg)
+	var flags activateOneContainerStateMachineFlags
+	switch {
+	case c.state.TotalActivatedContainers() == 0:
+		// If this is the first container to be activated, permit either
+		// a platform or recovery keyslot to be used.
+		flags = activatePermitPlatformKey | activatePermitRecoveryKey
+	case c.state.TotalActivatedContainers() == 1 && c.state.NumActivatedContainersWithRecoveryKey() == 1:
+		// If this is the second container to be activated and the first
+		// was activated with a recovery keyslot, permit either a platform
+		// or recovery keyslot to be used. The platform keyslot must be
+		// protected with a platform registered with the
+		// PlatformProtectedByStorageContainer flag.
+		flags = activatePermitPlatformKey | activatePermitRecoveryKey | activateRequirePlatformKeyProtectedByStorageContainer
+	case c.state.NumActivatedContainersWithRecoveryKey() > 1:
+		// If more than one storage container has been unlocked with a
+		// recovery key, then only permit recovery keyslots. We can only
+		// associate keyslots protected by a platform registered with the
+		// PlatformProtectedByStorageContainer flag with a single container,
+		// meaning that 1 or more storage containers might be supplied by
+		// an adversary.
```

Comment:
is this true in general? can't the first container contain multiple plainkeys?

Reply:
It can, but that's not the issue here. We can make use of the plainkey platform to associate 2 storage containers in the case where the source of the plainkey is unlocked using a recovery key, but the plainkey keyslots can only be associated with 1 source container.

Thinking about this again, if the user sat in front of the device knows the recovery key for 2 storage containers, then it's probably ok for me to remove this restriction.